### PR TITLE
Add Rolling Transform primitives

### LINKED ISSUE #1756: Add rolling gap primitives to Transform Primitives with integer parameters
As a user, I wish I could use Featuretools to engineer transform features that are generated over a rolling window within a single column. The window over which aggregations are calculated should be defined with two parameters: `window_length` and `gap`. These sorts of rolling aggregations can be useful for solving time series machine learning problems, which benefit from looking at previous observations. Therefore, adding these primitives will increase Featuretools' capabilities for time series feature engineering.

We should add primitives that can calculate rolling aggregations for the following operations: min, max, mean, std, count (?), and sum (?).

Window length and gap should be specified with integers representing the number of rows over which the gap and window should span. Note: For the best usage in time series problems, this assumes that the data has a uniform observation frequency (daily, hourly ,etc). This implementation will not enforce this uniform frequency.

Input types will require a Datetime time index to be passed in along with the column over which rolling aggregations will be calculated.

### PR DESCRIPTION
**Pull request by tamargrey**
Adds five rolling transform primitives that all takeinteger  `window_length`, `gap` and `min_periods` parameters.
closes #1756

### CODE REVIEW

File: featuretools/primitives/standard/rolling_transform_primitive.py
Comment (tamargrey) [inline featuretools/primitives/standard/rolling_transform_primitive.py:62]: Previously the default window-length was `1d`, which makes sense and will be a valid window with multiple instances present if the data's sampling frequency is less than 1 day. At this point in the implementation, though, we do not have access to date offsets, and setting the window length to `1` doesn't make much sense, as it would never end up with more than one instance present in the window. 

I think I might prefer using `window_length=7` or something that has some time-related meaning. But chose 3 for the moment just to show the arbitrary nature of having to set a default for this parameter. 

Also, when we implement #1757, we can go back to setting the default to be `1d`, but for now, I don't think there's a way to ensure this won't be a breaking change when that's implemented.

File: featuretools/primitives/utils.py
Comment (tamargrey) [inline featuretools/primitives/utils.py:265]: Created an issue to remove this eventually https://github.com/alteryx/featuretools/issues/1771

**Review by freddyaboulton — COMMENTED** [review]
@tamargrey The behavior of these primitives makes sense to me! The only thing that stood out to me was the difference in parameter definitions between EvalML and Featuretools that we chatted about. When we get to datestrings we will need your help in transforming the EvalML definitions to featuretools, but we agreed the "add 1" operation is possible. 

Don't think EvalML will ever set min_periods to be different than window_length but that may be useful for other kinds of problems. 

If users use `RollingSTD` with min_periods=1, there will be one extra nan compared to the other primitives because the standard deviation of a single observation is not defined. That makes sense to me.

File: featuretools/primitives/utils.py
Comment (freddyaboulton) [inline featuretools/primitives/utils.py:242]: nit-pick: The use of "skipped" confused me at first. If gap=1, and we're saying that 1 row prior to each instance will be skipped, that makes me think it's doing what gap=2 actually does.

**Review by thehomebrewnerd — COMMENTED** [review]
Overall this is looking decent, but just have a couple things:
- For the tests in `test_rolling_primitive.py` and `test_deep_feature_synthesis.py` we are only using the pandas version of the test fixture series. Should/could we also test with Dask/Koalas?
- Do we have any concerns with changing the init arguments for these primitives from their previous definitions? I think this means any existing applications where these were used as premium primitives will likely break, but I'm not sure that is an issue. Just wanted to mention in case it is important.

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py
Comment (tamargrey) [reply featuretools/tests/synthesis/test_deep_feature_synthesis.py]: yep, changed!

**Comment by tamargrey** [comment]
@thehomebrewnerd 
> For the tests in `test_rolling_primitive.py` and `test_deep_feature_synthesis.py` we are only using the pandas version of the test fixture series. Should/could we also test with Dask/Koalas?

Using all three was originally my plan, which was why I created all three fixtures, but when I started running into various issues (koalas' implementation of rolling sometimes has different behavior than pandas, and dask was having other unexplained failutes), I ended up decided to just allow Pandas, which was how they previously were. I still left the dask and koalas fixtures in place, since `_roll_series_with_gap` is still tested with them, so in theory we could add the dask and koalas compatibility if it's requested a bit easier. 

Happy to remove the extra fixtures if it doesn't make sense to keep them in if the primitives aren't yet compatible. Or, if we're okay with this taking a little longer to get in, I could investigate the dask and koalas issues  further to try and get the primitives compatible. @gsheni may have thoughts on whether it'd be worth it.

> Do we have any concerns with changing the init arguments for these primitives from their previous definitions? I think this means any existing applications where these were used as premium primitives will likely break, but I'm not sure that is an issue. Just wanted to mention in case it is important.

We'd need to get the date offset implementation in at the same time to have the primitives not break if they are currently in use anywhere. That would be a bigger ask for this PR, that I think we didn't want to have to wait for.  Also ran this by @gsheni a while back to confirm this was okay. Can revisit if it's going to be problematic to break them.

**Comment by gsheni** [comment]
@tamargrey is correct. We are punting on the Koalas/Dask implementation of these primitives. Additionally, we are okay with this being a breaking change for the RollingX primitives (compared to premium primitives)

**Comment by thehomebrewnerd** [comment]
I think my vote would be to remove the Dask and Koalas fixtures if we are not supporting those types of inputs now, and we can add them back in when they are needed. Otherwise the could end up hanging around for a long time without really being needed.

File: featuretools/primitives/standard/rolling_transform_primitive.py
Comment (rwedge) [inline featuretools/primitives/standard/rolling_transform_primitive.py:27]: How's this sound

> The number of rows backward from the target instance to be skipped before the window of usable data begins.  Defaults to 0, which will include the target instance in the window

If `include_cutoff_time` is False does gap=0 function the same as gap=1 would normally?

Comment (tamargrey) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: I like your suggestion. The one thing I might change is just completely getting rid of the word "skipped" based off of @freddyaboulton's comment above. How about

> The number of rows backward from the target instance before the window of usable data begins. Defaults to 0, which will include the target instance in the window

I think just talking about "n rows back" better encapsulates the idea that one row back is immediately above and zero rows back is just that instance. Instead of also including the idea of "n rows skipped" which would lead me to think there would be `n` entire rows that are ignored before starting the window.

Comment (tamargrey) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: To your second point: unless I'm missing something, setting a cutoff time and having `include_cutoff_time=False` shouldn't have an impact on the values included in any window. 

In the example below where we have primitives set with gap=0 and gap=1 and it just impacts the final row.
image

If I set cutoff_times for each row to be its own time index value (and now we have to have include_cutoff_time=True or else we'll have no data even though windows technically would be before the cutoff time--may need to be a separate issue if we wanted to handle that case specifically), it has no impact at all. (If we also add `uses_full_dataframe = True` to the primitive because of that separate bug)

image

Comment (rwedge) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: Thanks for the explanation, I think I was confusing using the cutoff time to create a window with using the target instance for a window

Comment (tamargrey) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: Ahh fair. Anything I can add to the docstrings to help clarify what happens in these primitives?

Comment (rwedge) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: I think it's pretty clear

Comment (gsheni) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: We could put this information in our documentation. Add it as a section or a new page on Time Series. Use it as a chance to explain rolling primitives

Comment (tamargrey) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: Have an issue to do that (#1758 - could also just be adding an entirely new Time series page). Was imagining it could happen once we have a lot of the time series stuff in, but we could always start it now and update it as we go along

Comment (gsheni) [reply featuretools/primitives/standard/rolling_transform_primitive.py:27]: @tamargrey your call. happy with either.

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (tamargrey) [reply featuretools/tests/primitive_tests/test_primitive_utils.py]: removed the koalas fixture

File: featuretools/primitives/standard/rolling_transform_primitive.py
Comment (tamargrey) [reply featuretools/primitives/standard/rolling_transform_primitive.py]: Gonna leave the documentation to later. Should be a bit easier to create a more cohesive narrative when everything is implemented.

**Comment by rwedge** [comment]
@tamargrey can you pull in the latest changes?

File: featuretools/tests/conftest.py
Comment (rwedge) [inline featuretools/tests/conftest.py:571]: what if we kept the "rolling_series_pd" name to avoid having to check the definition to see if its a pandas only fixture or a parameterized one?

Comment (tamargrey) [reply featuretools/tests/conftest.py:571]: sure, happy to keep as `rolling_series_pd`

**Review by rwedge — DISMISSED** [review]
LGTM assuming test pass

**Review by gsheni — DISMISSED** [review]

File: featuretools/primitives/utils.py
Comment (thehomebrewnerd) [inline featuretools/primitives/utils.py:255]: Spelling error: `wnd`

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_primitive_utils.py:195]: Leftover `to_pandas` call?

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_primitive_utils.py:206]: Remove `to_pandas` call?

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_primitive_utils.py:231]: Remove `to_pandas` calls?

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_primitive_utils.py:221]: Any need to also test these with an input series that contains null values to confirm they are handled as expected?

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py
Comment (thehomebrewnerd) [inline featuretools/tests/synthesis/test_deep_feature_synthesis.py:417]: Might be overkill, but you could create a loop here and test all the new primitives, except maybe count which has a different naming scheme. That probably already gets tested elsewhere though.

File: featuretools/tests/primitive_tests/test_rolling_primitive.py
Comment (thehomebrewnerd) [inline featuretools/tests/primitive_tests/test_rolling_primitive.py:18]: Wondering just a little bit about the robustness of these tests. If I'm looking at this correctly we are more or less duplicating the primitive function here and confirming that the actual primitive function returns the same value as the version we have in the test. If we did something to change `_roll_series_with_gap` that causes incorrect results, that would also get propagated here and the test would pass, but hopefully be caught in the tests specific to that function.

Not sure if it is better to hard code the expected values or not? Not a blocker for me one way or the other, just something to consider.

Comment (tamargrey) [reply featuretools/tests/primitive_tests/test_rolling_primitive.py:18]: Yeah, my goal in the tests for this feature would be to catch any wrong or unexpected rolling behavior in the tests specific to `_roll_series_with_gap`. 

My goal with the primitive tests was to confirm that they behave exactly like `_roll_series_with_gap` so we could catch any differing behavior that could get introduced within the primitive classes themselves. Or, in the case of `RollingCount` where we're specifically making changes to the output of `_roll_series_with_gap`, test just that differing behavior.

With regard to whether we should hardcode the values, my hesitation with doing that here would just be in the fact that it can get confusing to update hardcoded values if/when things do change. I've had situations when I've broken hardcoded tests in a repo I wasn't super familiar with, and it was a pretty big struggle to figure out whether they broke because the hardcoded value should now just be updated, or if it was because my changes had unintended side effects. I have less of a problem with hardcoded values in the `_roll_series_with_gap` tests, though, so if I should be doing a better job of that, happy to make changes.

File: featuretools/primitives/utils.py
Comment (tamargrey) [reply featuretools/primitives/utils.py:255]: thanks for catching!

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (tamargrey) [reply featuretools/tests/primitive_tests/test_primitive_utils.py:195]: yep, will remove all the to_pandas calls in this file!

File: featuretools/tests/primitive_tests/test_primitive_utils.py
Comment (tamargrey) [reply featuretools/tests/primitive_tests/test_primitive_utils.py:221]: yep, good idea--will add

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py
Comment (tamargrey) [reply featuretools/tests/synthesis/test_deep_feature_synthesis.py:417]: yeah, the test below `test_make_rolling_count_off_datetime_feature` will cover count. Happy to iterate over all the rolling primitives

Comment (tamargrey) [reply featuretools/tests/synthesis/test_deep_feature_synthesis.py]: ended up parameterizing the various primitives instead of iterating over them

Comment (thehomebrewnerd) [reply featuretools/tests/synthesis/test_deep_feature_synthesis.py]: Even better!

File: featuretools/tests/primitive_tests/test_rolling_primitive.py
Comment (thehomebrewnerd) [reply featuretools/tests/primitive_tests/test_rolling_primitive.py:18]: Sounds good to me. Know need to make any changes.

**Review by thehomebrewnerd — APPROVED** [review]
LGTM. Seems like the test coverage CI failure is due to some flaky Dask behavior and nothing in this MR.

File: featuretools/utils/plot_utils.py
Comment (rwedge) [inline featuretools/utils/plot_utils.py:41]: try using graphviz.FORMATS

**Review by gsheni — APPROVED** [review]

**Pull request merged**
